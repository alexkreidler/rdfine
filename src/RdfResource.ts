/* eslint-disable camelcase,@typescript-eslint/camelcase */
import { defaultGraph } from '@rdfjs/data-model'
import { NamedNode, DatasetCore, BlankNode, DefaultGraph, Quad_Graph } from 'rdf-js'
import cf, { SafeClownface, SingleContextClownface } from 'clownface'
import ResourceFactoryImpl, { Constructor, Mixin, ResourceFactory, ResourceIndexer } from './lib/ResourceFactory'
import once from 'once'
import TypeCollectionCtor, { TypeCollection } from './lib/TypeCollection'

type ObjectOrFactory<T> = T | ((self: RdfResource) => T)

export type ResourceIdentifier = BlankNode | NamedNode

export interface RdfResource<D extends DatasetCore = DatasetCore> {
  readonly id: ResourceIdentifier
  readonly types: TypeCollection<D>
  readonly _selfGraph: SingleContextClownface<D, ResourceIdentifier>
  readonly _unionGraph: SingleContextClownface<D, ResourceIdentifier>
  readonly _graphId: Quad_Graph
  hasType (type: string | NamedNode): boolean
  _create<T extends RdfResource>(term: SingleContextClownface<D>, mixins?: Mixin<any>[] | [Constructor, ...Mixin<any>[]]): T & ResourceIndexer
}

export default class RdfResourceImpl<D extends DatasetCore = DatasetCore> implements RdfResource<D> {
  public readonly _selfGraph: SingleContextClownface<D, ResourceIdentifier>
  public readonly _unionGraph: SingleContextClownface<D, ResourceIdentifier>
  public readonly __initialized: boolean = false
  private readonly __initializeProperties: (() => boolean)
  public static __ns?: any
  public static factory: ResourceFactory = new ResourceFactoryImpl(RdfResourceImpl)

  public constructor(graph: SafeClownface<D> | { dataset: D; term: ResourceIdentifier; graph?: NamedNode | DefaultGraph }) {
    let selfGraph: SingleContextClownface<D, ResourceIdentifier>

    if ('_context' in graph) {
      const [{ term }, ...rest] = graph.toArray()

      if (rest.length > 0) {
        throw new Error(`RdfResource can only be initialized from a single node. Got ${rest.length} additional values.`)
      }

      if (term.termType !== 'BlankNode' && term.termType !== 'NamedNode') {
        throw new Error(`RdfResource cannot be initialized from a ${term.termType} node`)
      }

      selfGraph = graph.node(term)
    } else {
      selfGraph = cf(graph)
    }

    if (selfGraph._context[0].graph) {
      this._selfGraph = selfGraph
      this._unionGraph = cf({ dataset: selfGraph.dataset, term: selfGraph.term, graph: undefined })
    } else {
      this._selfGraph = cf({ dataset: selfGraph.dataset, term: selfGraph.term, graph: defaultGraph() })
      this._unionGraph = cf({ dataset: selfGraph.dataset, term: selfGraph.term })
    }

    this.types = new TypeCollectionCtor(this)

    this.__initializeProperties = once(() => {
      const self = this as any
      const defaults: Map<string, ObjectOrFactory<any>> = self.constructor.__defaults || new Map<string, ObjectOrFactory<any>>()
      defaults.forEach((value, key) => {
        const currentValue = self[key]
        const valueIsEmptyArray = Array.isArray(currentValue) && currentValue.length === 0
        const valueIsUndefined = typeof currentValue === 'undefined'

        if (!(valueIsEmptyArray || valueIsUndefined)) {
          return
        }

        if (typeof value === 'function') {
          self[key] = value(self)
          return
        }

        self[key] = value
      })

      return true
    })

    this.__initialized = this.__initializeProperties()
  }

  public get id(): ResourceIdentifier {
    return this._selfGraph.term
  }

  public get _graphId(): Quad_Graph {
    return this._selfGraph._context[0].graph!
  }

  public readonly types: TypeCollection<D>

  public hasType(type: string | NamedNode): boolean {
    return this.types.has(type)
  }

  public equals(other: RdfResource | undefined | null): boolean {
    if (!other) {
      return false
    }

    return this.id.equals(other.id)
  }

  public _create<T extends RdfResource>(term: SingleContextClownface<D>, mixins?: Mixin<any>[] | [Constructor, ...Mixin<any>[]]): T & ResourceIndexer {
    return (this.constructor as Constructor).factory.createEntity<T>(term, mixins)
  }
}
