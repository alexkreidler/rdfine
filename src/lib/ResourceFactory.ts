import RdfResourceImpl, { RdfResource } from './RdfResource'
import { Clownface } from 'clownface'

export type AnyFunction<A = any> = (...input: any[]) => A
export type Constructor<A extends RdfResource = RdfResource> = new (...input: any[]) => A
interface ShouldApply {
  shouldApply: boolean | ((entity: RdfResource) => boolean)
}

export type Mixin<T extends AnyFunction> = InstanceType<ReturnType<T>>

export class ResourceFactory<T extends AnyFunction> {
  private __mixins: Set<Mixin<any>> = new Set()
  public BaseClass: Constructor = RdfResourceImpl

  public addMixin(mixin: Mixin<T> & ShouldApply): void {
    this.__mixins.add(mixin)
  }

  public createEntity<R extends RdfResource>(term: Clownface, explicitMixins: Mixin<T>[] = []): RdfResource & R {
    const entity = new this.BaseClass(term)

    const mixins = [...this.__mixins].reduce<Mixin<T>[]>((selected, next: Mixin<T> & ShouldApply) => {
      if (next.shouldApply === true || (typeof next.shouldApply === 'function' && next.shouldApply(entity))) {
        if (!selected.includes(next)) {
          selected.push(next)
        }
      }

      return selected
    }, [...explicitMixins])

    const Type = mixins.reduce<Constructor>((Mixed: Constructor, Next: Mixin<T>) => Next(Mixed), this.BaseClass)
    ;(Type as any).__mixins = mixins

    return new Type(term) as RdfResource & R
  }
}

export const factory = new ResourceFactory()
