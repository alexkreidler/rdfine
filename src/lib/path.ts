import { NamedNode } from 'rdf-js'
import { namedNode } from 'rdf-data-model'
import RdfResourceImpl from './RdfResource'

export type PropRef = string | NamedNode

function getPredicate(resource: RdfResourceImpl, name: PropertyKey): NamedNode {
  return (resource.constructor as any).__ns[name.toString()]
}

function predicate(termOrString: PropRef): NamedNode {
  if (typeof termOrString === 'string') {
    return namedNode(termOrString)
  } else {
    return termOrString
  }
}

export function getPath(protoOrDescriptor: any, name: PropertyKey, path?: PropRef | PropRef[]) {
  return (path ? Array.isArray(path) ? path : [path] : [getPredicate(protoOrDescriptor, name)])
    .map(termOrString => predicate(termOrString))
}
