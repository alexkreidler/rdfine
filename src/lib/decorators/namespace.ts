import ns from '@rdfjs/namespace'
import { RdfResource } from '../RdfResource'

type Constructor<T = RdfResource> = {
  new (...args: any[]): T
  __ns?: any
};

export function namespace(stringOrNamespace: string | unknown) {
  return <T extends RdfResource>(classOrDescriptor: Constructor<T>) => {
    classOrDescriptor.__ns = typeof stringOrNamespace === 'string'
      ? ns(stringOrNamespace) : ns
  }
}
