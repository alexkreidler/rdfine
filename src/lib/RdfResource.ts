import { Term, NamedNode } from 'rdf-js'
import Clownface from 'clownface/lib/Clownface'
import ns from '@rdfjs/namespace'
import { ResourceFactory } from './ResourceFactory'

const rdf = ns('http://www.w3.org/1999/02/22-rdf-syntax-ns#')

export interface RdfResource {
  readonly id: Term | null
  readonly types: Term[]
  hasType (type: string | NamedNode): boolean
}

export default class RdfResourceImpl implements RdfResource {
  public _node: Clownface

  public static _factory = new ResourceFactory(RdfResourceImpl)

  public constructor(cf: Clownface) {
    this._node = cf
  }

  public get id() {
    return this._node.term
  }

  public get types() {
    return this._node.out(rdf.type).terms
  }

  public hasType(type: string | NamedNode) {
    const typeNode = typeof type === 'string' ? this._node.namedNode(type) : type

    return this._node.has(rdf.type, typeNode).terms.length > 0
  }
}
