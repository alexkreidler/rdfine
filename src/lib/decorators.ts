import { Literal, NamedNode, Term } from 'rdf-js'
import Clownface from 'clownface/lib/Clownface'
import rdf from 'rdf-data-model'
import { Mixin } from './ResourceFactory'
import RdfResource from './RdfResource'

const trueLiteral: Literal = rdf.literal('true', rdf.namedNode('http://www.w3.org/2001/XMLSchema#boolean'))

type PropRef = string | NamedNode

interface AccessorOptions {
  path?: PropRef | PropRef[]
  array?: boolean
}

interface ResourceAccessorOptions extends AccessorOptions {
  as?: Mixin<any>[]
}

type Constructor<T = RdfResource> = {
  new (...args: any[]): T
  __ns?: any
};

interface LiteralAccessorOption extends AccessorOptions {
  type?: typeof Boolean
}

function predicate(cf: Clownface, termOrString: PropRef) {
  if (typeof termOrString === 'string') {
    return cf.namedNode(termOrString)
  } else {
    return cf.node(termOrString)
  }
}

function getNode(cf: Clownface, path: Clownface[]) {
  return path
    .reduce((node, prop) => {
      return node.out(prop)
    }, cf)
}

function getPredicate(cf: any, name: PropertyKey): NamedNode {
  return (cf.constructor).__ns[name.toString()]
}

function getPath(protoOrDescriptor: any, cf: Clownface, name: PropertyKey, path?: PropRef | PropRef[]) {
  return (path ? Array.isArray(path) ? path : [path] : [getPredicate(protoOrDescriptor, name)])
    .map(termOrString => predicate(cf, termOrString))
}

export function resource(options: ResourceAccessorOptions = {}) {
  return (protoOrDescriptor: any, name: PropertyKey): any => {
    Object.defineProperty(protoOrDescriptor, name, {
      get(this: any): any {
        const node = getNode(this._node, getPath(protoOrDescriptor, this._node, name, options.path))

        const values = node.map(term => {
          return this.constructor._factory.createEntity(term, options.as)
        })

        if (options.array === true) {
          return values
        }

        if (values.length > 1) {
          throw new Error('Multiple terms found where 0..1 was expected')
        }

        return values[0]
      },

      set(this: RdfResource, value: any) {
        const path = getPath(protoOrDescriptor, this._node, name, options.path)
        const node = path.length === 1 ? this._node : getNode(this._node, path.slice(path.length - 1))

        const lastPredicate = path[path.length - 1]
        node.deleteOut(lastPredicate)
          .addOut(lastPredicate, value)
      },
    })
  }
}

export function term(options: AccessorOptions = {}) {
  return (protoOrDescriptor: any, name: PropertyKey): any => {
    Object.defineProperty(protoOrDescriptor, name, {
      get(this: any): any {
        const node = getNode(this._node, getPath(protoOrDescriptor, this._node, name, options.path))

        const values = node.terms

        if (options.array === true) {
          return values
        }

        if (values.length > 1) {
          throw new Error('Multiple terms found where 0..1 was expected')
        }

        return values[0]
      },

      set(this: RdfResource, value: Term) {
        const path = getPath(protoOrDescriptor, this._node, name, options.path)
        const node = path.length === 1 ? this._node : getNode(this._node, path.slice(path.length - 1))

        const lastPredicate = path[path.length - 1]
        node.deleteOut(lastPredicate)
          .addOut(lastPredicate, value)
      },
    })
  }
}

export function literal(options: LiteralAccessorOption = {}) {
  return (protoOrDescriptor: any, name: PropertyKey): any => {
    Object.defineProperty(protoOrDescriptor, name, {
      get(this: RdfResource): any {
        const node = getNode(this._node, getPath(protoOrDescriptor, this._node, name, options.path))

        if (options.type === Boolean) {
          return trueLiteral.equals(node.term as any) // TODO: fix equals typing
        }

        return node.value
      },

      set(this: RdfResource, value: any) {
        const path = getPath(protoOrDescriptor, this._node, name, options.path)
        const node = path.length === 1 ? this._node : getNode(this._node, path.slice(path.length - 1))

        const lastPredicate = path[path.length - 1]
        node.deleteOut(lastPredicate)
          .addOut(lastPredicate, value)
      },
    })
  }
}

export function namespace(ns: any) {
  return <T extends RdfResource>(classOrDescriptor: Constructor<T>) => {
    classOrDescriptor.__ns = ns
  }
}
