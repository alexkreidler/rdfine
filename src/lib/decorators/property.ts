import { DatasetCore, NamedNode, Term } from 'rdf-js'
import RdfResourceImpl, { RdfResource } from '../RdfResource'
import { getPath, PropRef } from '../path'
import cf, { SafeClownface, SingleContextClownface } from 'clownface'
import { isList, enumerateList } from '../rdf-list'
import { ClassElement } from './index'
import literalPropertyDecorator from './property/literal'
import resourcePropertyDecorator from './property/resource'
import { rdf } from '../vocabs'

export interface AccessorOptions {
  values?: 'array' | 'list'
  path?: PropRef | PropRef[]
  strict?: true
  namedGraphs?: Partial<NamedGraphsOptions>
}

interface NamedGraphsOptions {
  combineSubjects: boolean
  crossBoundaries: boolean
}

function getNode(node: SingleContextClownface, path: NamedNode[]): SafeClownface {
  return path
    .reduce<SafeClownface>((node, prop) => {
    return node.out(prop)
  }, node)
}

function onlyUnique <T>(areEqual: (left: T, right: T) => boolean) {
  return function (value: T | T[], index: number, self: Array<T | T[]>): boolean {
    if (Array.isArray(value)) {
      return true
    }

    const found = self.findIndex(other => {
      if (Array.isArray(other)) {
        return true
      }

      return areEqual(value, other)
    })

    return found === index
  }
}

type ArrayOrSingle<T> = T | T[]
export type ObjectOrFactory<TSelf, T, TTerm extends Term> =
  ArrayOrSingle<T | TTerm | SingleContextClownface<DatasetCore, TTerm>> |
  ((self: TSelf) => ArrayOrSingle<T | TTerm | SingleContextClownface<DatasetCore, TTerm>>)

interface PropertyDecoratorOptions<T extends RdfResource, TValue, TTerm extends Term> extends AccessorOptions {
  fromTerm: (this: T, obj: SingleContextClownface) => TValue
  toTerm: (value: TValue) => TTerm
  assertSetValue: (value: RdfResource | Term | SingleContextClownface | TValue) => boolean
  valueTypeName: string
  initial?: ObjectOrFactory<T, TValue, TTerm>
  compare: (left: TValue, right: TValue) => boolean
}

function createProperty<T extends RdfResource, TValue, TTerm extends Term>(proto: any, name: string, options: PropertyDecoratorOptions<T, TValue, TTerm>) {
  const { path, fromTerm, toTerm, assertSetValue, valueTypeName, initial, strict, compare } = options
  const { combineSubjects = false, crossBoundaries = false } = options.namedGraphs || {}
  const values = options.values || (crossBoundaries || combineSubjects ? 'array' : 'single')

  if (values === 'list' && crossBoundaries) {
    throw new Error('RDF Lists are not supported with property operating across multiple named graphs')
  }

  Object.defineProperty(proto, name, {
    get(this: T & RdfResourceImpl): unknown {
      const rootNode = combineSubjects ? this._unionGraph : this._selfGraph
      const pathNodes = getPath(proto, name, path)
      const nodes = getNode(rootNode, pathNodes)

      const objectNodes = nodes.toArray().reduce((current, node) => {
        if (crossBoundaries) {
          const quadsWithSubject = rootNode.dataset.match(node.term)
          const quadsWithObject = rootNode.dataset.match(null, null, node.term)
          const allQuads = [...quadsWithSubject, ...quadsWithObject]

          const allNodes = allQuads.reduce((graphNodes, quad) => {
            if (!graphNodes.has(quad.graph.value)) {
              graphNodes.set(quad.graph.value, cf({
                dataset: this._unionGraph.dataset,
                term: quad.subject,
                graph: quad.graph,
              }))
            }

            return graphNodes
          }, new Map<string, SingleContextClownface>())
          return [...current, ...allNodes.values()]
        }

        return [...current, node]
      }, [] as SingleContextClownface[])

      const returnValues = objectNodes.map((obj, index) => {
        if (isList(obj)) {
          if (index > 0) {
            throw new Error('Lists of lists are not supported')
          }

          return enumerateList(this, obj, fromTerm.bind(this))
        }

        return fromTerm.call(this, obj)
      }).filter(onlyUnique(compare))

      if (values === 'array') {
        return returnValues
      }

      if (values === 'list') {
        return returnValues[0] || []
      }

      if (returnValues.length > 1) {
        throw new Error(`${name}: Multiple terms found where 0..1 was expected`)
      }

      if (Array.isArray(returnValues[0])) {
        throw new Error(`${name}: RDF List found where 0..1 object was expected`)
      }

      if (this.__initialized && strict && returnValues.length === 0) {
        throw new Error(`Object not found for property ${name}`)
      }

      return returnValues[0]
    },

    set(this: RdfResourceImpl, value: ArrayOrSingle<RdfResource | Term | SingleContextClownface>) {
      if (values === 'single' && Array.isArray(value)) {
        throw new Error(`${name}: Cannot set array to a non-array property`)
      }

      const pathNodes = getPath(proto, name, path)
      const subject = pathNodes.length === 1
        ? this._selfGraph
        : getNode(this._selfGraph, pathNodes.slice(0, pathNodes.length - 1))

      const lastPredicate = pathNodes[pathNodes.length - 1]

      subject.out(lastPredicate).forEach(obj => {
        if (isList(obj)) {
          subject.deleteList(lastPredicate)
        }
      })

      subject.deleteOut(lastPredicate)

      if (value === null || typeof value === 'undefined') {
        return
      }

      let valueArray: Array<RdfResource | Term | SingleContextClownface | TValue>
      if (Array.isArray(value)) {
        valueArray = value
      } else {
        valueArray = [value]
      }

      const termsArray = valueArray.map(value => {
        if (!assertSetValue(value)) {
          const pathStr = pathNodes.map(p => `<${p}>`).join('/')
          throw new Error(`Unexpected value for path ${pathStr}. Expecting a ${valueTypeName} or RDF/JS term`)
        }

        if (typeof value === 'object' && 'termType' in value) {
          return value
        }

        if (typeof value === 'object' && 'term' in value) {
          return value.term
        }

        if (typeof value === 'object' && 'id' in value) {
          return value.id
        }

        return toTerm(value)
      })

      if (values === 'list') {
        if (termsArray.length === 0) {
          subject.addOut(lastPredicate, rdf.nil)
        } else {
          subject.addList(lastPredicate, termsArray)
        }
      } else {
        subject.addOut(lastPredicate, termsArray)
      }
    },
  })

  if (typeof initial !== 'undefined') {
    if (!Object.hasOwnProperty.call(proto.constructor, '__defaults')) {
      proto.constructor.__defaults = new Map()
    }

    proto.constructor.__defaults.set(name, initial)
  }
}

const legacyProperty =
  <T extends RdfResource, TValue, TInitial, TTerm extends Term>(options: PropertyDecoratorOptions<T, TValue, TTerm>, proto: Record<string, unknown>, name: PropertyKey) => {
    createProperty(proto, name.toString(), options)
  }

const standardProperty =
  <T extends RdfResource, TValue, TInitial, TTerm extends Term>(options: PropertyDecoratorOptions<T, TValue, TTerm>, element: ClassElement) => {
    return {
      kind: 'field',
      key: Symbol(),
      placement: 'own',
      descriptor: {},
      // When @babel/plugin-proposal-decorators implements initializers,
      // do this instead of the initializer below. See:
      // https://github.com/babel/babel/issues/9260 extras: [
      //   {
      //     kind: 'initializer',
      //     placement: 'own',
      //     initializer: descriptor.initializer,
      //   }
      // ],
      initializer(this: { [key: string]: unknown }) {
        if (typeof element.initializer === 'function') {
          this[element.key as string] = element.initializer.call(this)
        }
      },
      finisher(clazz: any) {
        createProperty(clazz.prototype, element.key.toString(), options)
      },
    }
  }

export function propertyDecorator<T extends RdfResource, TValue, TTerm extends Term>(options: PropertyDecoratorOptions<T, TValue, TTerm>) {
  return (protoOrDescriptor: RdfResource|ClassElement, name?: PropertyKey): any =>
    (name !== undefined)
      ? legacyProperty(options, protoOrDescriptor as any, name)
      : standardProperty(options, protoOrDescriptor as ClassElement)
}

interface TermOptions <TSelf>{
  initial?: ObjectOrFactory<TSelf, Term, Term>
}

export function property<R extends RdfResource>(options: AccessorOptions & TermOptions<R> = {}) {
  return propertyDecorator<R, Term, Term>({
    ...options,
    namedGraphs: {
      crossBoundaries: false,
      combineSubjects: options.namedGraphs?.combineSubjects,
    },
    values: options.namedGraphs?.crossBoundaries ? 'array' : options.values,
    fromTerm: (obj) => obj.term,
    toTerm: value => value,
    valueTypeName: 'RDF/JS term object',
    assertSetValue: () => true,
    compare: (left, right) => left && left.equals(right),
  })
}

property.literal = literalPropertyDecorator

property.resource = resourcePropertyDecorator
