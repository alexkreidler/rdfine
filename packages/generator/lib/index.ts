import { Clownface } from 'clownface'
import { NamespaceBuilder } from '@rdfjs/namespace'
import { Debugger } from 'debug'
import { NamedNode } from 'rdf-js'

type PropertyType = 'string' | 'number' | 'boolean' | 'Date' | NamedNode['termType']

export type TypeMap = Record<string, PropertyType>

export interface Context {
  vocabulary: Clownface
  namespace: NamespaceBuilder
  prefix: string
  defaultExport: string
  typeMappings: TypeMap
  excludedTypes: string[]
  log: {
    debug: Debugger
    warn: Debugger
    error: Debugger
  }
}
