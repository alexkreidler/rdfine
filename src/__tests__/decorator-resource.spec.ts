import { prefixes } from '@zazuko/rdf-vocabularies'
import { literal, resource, term } from '../lib/decorators'
import RdfResource from '../lib/RdfResource'
import { parse, vocabs } from './_helpers'
import { NamedNode } from 'rdf-js'

const { ex, foaf, schema } = vocabs

describe('decorator', () => {
  describe('resource', () => {
    it('returns a resource object using path predicate', async () => {
      // given
      const dataset = await parse(`
        @prefix ex: <${prefixes.ex}> .
        @prefix foaf: <${prefixes.foaf}> .
        
        ex:res foaf:friend ex:friend .
      `)

      class Resource extends RdfResource {
        @resource({ path: foaf.friend })
        friend?: RdfResource
      }

      // when
      const instance = new Resource({
        dataset,
        term: ex.res,
      })
      const friend = instance.friend

      // then
      expect(friend).toBeInstanceOf(RdfResource)
      expect(friend!.id.value).toEqual(ex.friend.value)
    })

    it('sets a relation between resources', async () => {
      // given
      const dataset = await parse(`
        @prefix ex: <${prefixes.ex}> .
        @prefix schema: <${prefixes.schema}> .
        
        ex:john a schema:Person .
        ex:jane a schema:Person .
      `)

      class Resource extends RdfResource {
        @resource({ path: schema.spouse })
        spouse?: RdfResource
      }
      const john = new Resource({ dataset, term: ex.john })
      const jane = new Resource({ dataset, term: ex.jane })

      // when
      john.spouse = jane

      // then
      expect(john._node.dataset.toCanonical()).toMatchSnapshot()
    })

    it('setting null removes relation', async () => {
      // given
      const dataset = await parse(`
        @prefix ex: <${prefixes.ex}> .
        @prefix schema: <${prefixes.schema}> .
        
        ex:john a schema:Person .
        ex:jane a schema:Person .
        
        ex:john schema:spouse ex:jane .
      `)

      class Resource extends RdfResource {
        @resource({ path: schema.spouse })
        spouse?: RdfResource | null
      }
      const john = new Resource({ dataset, term: ex.john })

      // when
      john.spouse = null

      // then
      expect(john._node.dataset.toCanonical()).toMatchSnapshot()
    })
  })

  describe('term', () => {
    it('returns raw term from property', async () => {
      // given
      const dataset = await parse(`
        @prefix ex: <${prefixes.ex}> .
        @prefix foaf: <${prefixes.foaf}> .
        
        ex:res foaf:friend ex:friend .
      `)
      class Resource extends RdfResource {
        @term({ path: foaf.friend })
        friend?: NamedNode
      }

      // when
      const instance = new Resource({
        dataset,
        term: ex.res,
      })
      const friend = instance.friend

      // then
      expect(friend!.value).toEqual(ex.friend.value)
    })
  })

  describe('literal', () => {
    it('returns string value of literal', async () => {
      // given
      const dataset = await parse(`
        @prefix ex: <${prefixes.ex}> .
        @prefix schema: <${prefixes.schema}> .
        
        ex:res schema:name "John Doe" .
      `)
      class Resource extends RdfResource {
        @literal({ path: schema.name })
        name?: string
      }

      // when
      const instance = new Resource({
        dataset,
        term: ex.res,
      })

      // then
      expect(instance.name).toEqual('John Doe')
    })

    it('returns boolean when type is set', async () => {
    // given
      const dataset = await parse(`
        @prefix ex: <${prefixes.ex}> .
        @prefix schema: <${prefixes.schema}> .
        @prefix xsd: <${prefixes.xsd}> .
        
        ex:res ex:isMarried true .
        ex:res ex:isSingle "false"^^xsd:boolean .
      `)
      class Resource extends RdfResource {
        @literal({ path: ex.isMarried, type: Boolean })
        married?: boolean

        @literal({ path: ex.isSingle, type: Boolean })
        single?: boolean
      }

      // when
      const instance = new Resource({
        dataset,
        term: ex.res,
      })

      // then
      expect(instance.single).toStrictEqual(false)
      expect(instance.married).toStrictEqual(true)
    })
  })
})
