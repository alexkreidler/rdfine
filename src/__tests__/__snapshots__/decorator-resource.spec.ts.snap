// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decorator resource sets a relation between resources 1`] = `
"<http://example.com/jane> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://schema.org/Person> .
<http://example.com/john> <http://schema.org/spouse> <http://example.com/jane> .
<http://example.com/john> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://schema.org/Person> .
"
`;

exports[`decorator resource setting null removes relation 1`] = `
"<http://example.com/jane> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://schema.org/Person> .
<http://example.com/john> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://schema.org/Person> .
"
`;
